ignitionrobotics/ign-gazebo
# Add parameters system

# 🎉 New feature

Connected to https://github.com/ignitionrobotics/ign-msgs/pull/211.
Hints of how to review the PR: https://github.com/ignitionrobotics/ign-gazebo/pull/1280#issuecomment-1010421915.

## Summary

This PR adds:

- A parameters system:
  - Other systems can declare parameter using the `ignition::gazebo::DeclareParameter()` helper function.
  - Parameters can be introspected or set from external processes through three service provided by the `Parameters` system:
    - `/world/<world_name>/list_parameters` service: List available parameters names and types.
    - `/world/<world_name>/get_parameter` service: Get the type and value of a parameter.
    - `/world/<world_name>/set_parameter` service: Set a parameter: parameter name, value and type need to be provided.
  - Parameters types are protobuf messages.
  - Parameters are automatically associated to a component in the entity component manager, there's no separate "register" of parameters. When the parameter is set, the associated component is updated.
    The component data type must be a protobuf message to be declared as a parameter.
- A `ign param` command line tool. A simple cli tool to interact with parameters.
  Though interacting with parameters using `ign service` is possible, it's not as easy as parameter values [are serialized](https://github.com/ignitionrobotics/ign-msgs/blob/7086fe5c3546a4f086ed77cc468edac56397a1a3/proto/ignition/msgs/parameter.proto#L33).

Summary of what's needed to declare and use a parameter can be seen in the wheel slip demo [branch](https://github.com/ignitionrobotics/ign-gazebo/tree/ivanpauno/parameters-impl-draft-wheel-slip-demo): [diff](https://github.com/ignitionrobotics/ign-gazebo/commit/1d2011348ba59c62bdac58a7a1a9cb4b64bab540).

## Test it


I haven't written automated tests yet, but there's a demo using parameters in the wheel slip plugin:

- https://github.com/ignitionrobotics/ign-gazebo/tree/ivanpauno/parameters-impl-draft-wheel-slip-demo
- https://github.com/ignitionrobotics/ign-msgs/tree/ivanpauno/parameters-impl-draft-wheel-slip-demo

Steps to run the demo:

- Build from source ignition `edifice` checking out the two branches specified above (I can provide PRs for another version if needed).
- Download the following sdf demo world: https://gist.github.com/ivanpauno/2d7ad7bb4ea8ca88aaaa89458999cb0f
  ```
  wget https://gist.githubusercontent.com/ivanpauno/2d7ad7bb4ea8ca88aaaa89458999cb0f/raw/3710fb389f9c841e4992835a4dc1dd5ab933cefd/wheel_slip_parameters_world.xml
  ```
- Launch the downloaded world in ignition gazebo, in a terminal with the previous workspace sourced.
  ```
  ign gazebo ./wheel_slip_parameters_world.xml
  ```
- In another terminal with the workspace sourced:
  - List available parameters:
    ```
    ign param -l
    ```
    <details>
      <summary>Expected output</summary>

      > systems.wheel_slip.trisphere_cycle1.wheel_front            [ignition.msgs.WheelSlipParameters]
      > systems.wheel_slip.trisphere_cycle1.wheel_rear_right            [ignition.msgs.WheelSlipParameters]
      > systems.wheel_slip.trisphere_cycle0.wheel_rear_right            [ignition.msgs.WheelSlipParameters]
      > systems.wheel_slip.trisphere_cycle1.wheel_rear_left            [ignition.msgs.WheelSlipParameters]
      > systems.wheel_slip.trisphere_cycle0.wheel_rear_left            [ignition.msgs.WheelSlipParameters]
      > systems.wheel_slip.trisphere_cycle0.wheel_front            [ignition.msgs.WheelSlipParameters]

    </details>
  - Get the `systems.wheel_slip.trisphere_cycle1.wheel_front` parameter:
    ```
    ign param --get -n systems.wheel_slip.trisphere_cycle1.wheel_front
    ```
    <details>
      <summary>Expected output</summary>

      >
      > Getting parameter [systems.wheel_slip.trisphere_cycle1.wheel_front] for world [wheel_slip]...
      > Parameter type [ignition.msgs.WheelSlipParameters]
      > 
      > \------------------------------------------------
      > slip_compliance_lateral: 1
      > slip_compliance_longitudinal: 1
      > \------------------------------------------------

    </details>
  - Set the `systems.wheel_slip.trisphere_cycle1.wheel_front` parameter to a different value:
    ```
    ign param --set -n systems.wheel_slip.trisphere_cycle1.wheel_front -t ignition.msgs.WheelSlipParameters -m "
      slip_compliance_lateral: 1
      slip_compliance_longitudinal: 2"
    ```
    <details>
      <summary>Expected output</summary>

      >
      > Setting parameter [systems.wheel_slip.trisphere_cycle1.wheel_front] for world [wheel_slip]...
      >
      > Parameter successfully set!

    </details>
  - Get the `systems.wheel_slip.trisphere_cycle1.wheel_front` parameter again, its value should have changed:
    ```
    ign param --get -n systems.wheel_slip.trisphere_cycle1.wheel_front
    ```
    <details>
      <summary>Expected output</summary>

      >
      > Getting parameter [systems.wheel_slip.trisphere_cycle1.wheel_front] for world [wheel_slip]...
      > Parameter type [ignition.msgs.WheelSlipParameters]
      > 
      > \------------------------------------------------
      > slip_compliance_lateral: 1
      > slip_compliance_longitudinal: 2
      > \------------------------------------------------

## Checklist

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.
